zirgen/docs/05_Muxes.md

```diff
@@ -0,0 +1,187 @@
+might make this look a bit more familiar. This code is exactly the same as the
+previous example:
+
+```
+component Top() {
+  first := NondetReg(IsFirstCycle());
+
+  if (first) {
+    Init()
+  } else {
+    Step()
+  }
+}
+```
+
+## Super Components
+
+We mentioned just a moment ago that muxes are expressions, which raises an
+interesting question: what is the "type" of a mux expression? Clearly, if all
+the arms have the same types, the mux should be able to be used as that type. It
+would be quite restrictive to forbid this, and indeed the following logs either
+"a is 5" or "a is 7" depending on the value of `cond`.
+
+```
+component Top() {
+  ...
+  a := if (cond) { 5 } else { 7 };
+  Log("a is %u", a);
+}
+```
+
+Consider instead a more complicated case like the following, where the mux arms
+have the slightly different types `NondetReg` and `Val`. Well, a `NondetReg` is
+already very similar to a `Val` in that you can apply field operations to it,
+pass it to constructors that expect a `Val`, and use it every other way that you
+can use a `Val`.
+
+```
+component Top() {
+  ...
+  a := if (cond) { NondetReg(GetCycle()) } else { 0 };
+  Log("a is %u", a);
+}
+```
+
+The reason that this works is that a `NondetReg` can be converted to a `Val`
+through its super chain. Since a mux must have exactly one active arm, but in
+general any of them may be active depending on the row of the trace, the mux
+should be convertible to any type that all the arms can be — and since a type is
+convertible to any type in its super chain, a mux's super type is the _least
+common super type_ of its arms.
+
+The least common super type of a set of components is the first type that occurs
+in the super chain of all of the members of the set. To give some examples,
+consider this type hierarchy:
+
+```mermaid
+graph TD;
+    Component --- A
+    A --- B
+    B --- D
+    B --- E
+    A --- C
+    C --- F
+    C --- G
+```
+
+The following are all true statements:
+* The least common super of {`B`} is `B`
+* The least common super of {`B`, `C`} is `A`
+* The least common super of {`D`, `E`} is `B`
+* The least common super of {`B`, `D`} is `B`
+* The least common super of {`B`, `F`} is `A`
+
+One last note on the topic of least common supers: the word "least" is chosen
+because it is the common super type that appears "lowest" in the type hierarchy
+diagram. This ties to the fact that algebraically the set of component types
+form a semilattice with the least common super as the join operation.
+
+## Layout
+
+The layout of a mux in practice is the layout of its active arm, but it's
+advantageous to give it some additional structure. Since the type of a mux
+expression is the least common super of all its arms, it is often useful to
+refer to its value in previous cycles. In order for this to work naturally, the
+least common super must have the same layout in all arms, which is guaranteed by
+the language and handled automatically by the compiler.
```

Comment:
This sounds like a little too much in the way of implementation details.  It reads to me like `we handle this case specially internally in order for it to work like you'd expect`.

Reply:
ya hard for me to gauge but i have been finding it helpful elsewhere to aim for a level of detail that means the docs will still be accurate in 3/6 months.

Reply:
Sure, it seems important to document where the language guarantees shared layout, but I could probably pare down the practical justification for the decision. I think it's unlikely that this guarantee will go away, since all of our circuits depend on it heavily.